### DISCUSSION TITLE
Swagger-Native Read Only or Write Only Fields in FastAPI

### DISCUSSION THREAD

**Discussion by avivgood** [fastapi/fastapi#12233] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel, Field
from fastapi import FastAPI
import uvicorn
from uuid import UUID

app = FastAPI()


class ExampleModel(BaseModel):
    uid: UUID = Field(read_only=True)
    field: str


@app.post("/test/")
def test(model: ExampleModel):
    pass



@app.get("/test/")
def test() -> ExampleModel:
    pass


uvicorn.run(app)
```

### Description

I was reading the [swagger specification for objects](https://swagger.io/docs/specification/data-models/data-types/#object) and I noticed a feature called "Read-Only and Write-Only Properties". It allows the same Model for GET and POST, even with slight variations. A very common one is that we don't want to receive ID in the POST as we create a new object, but we do want to receive it in the GET response. [In such cases](https://github.com/fastapi/fastapi/discussions/6243) what is usually done by having a base model and then inheriting from it to two (or more) separate models: for example one for POST and one for GET. I find this solution pretty clunky as it drastically increases the number of models one needs to maintain. Also, it means a lot of similar models appear as different models in the schemas section. It also usually leads to awkward naming such as `FooPost` and `FooGet`. I think it would be a lot more neat if we were to have read / write-only fields as supported natively by Swagger.  AFAIK this does not exist yet. Does something like what is shown in the attached code exist in FastApi? If not I think it would be nice for it to be added.

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.10.12

**Comment by YuriiMotov** [comment]
Interesting feature.
But this should be implemented on Pydantic's side first..

**Reply by avivgood** [reply]
I don't believe this has anything to do with Pydantic. As far as Pydantic is concerned, it is not running as part of a REST API and is not used to generate an OpenAPI specification. This should, IMO, be implemented only in the FastAPI side. AFAIK you can add custom kwargs to fields constructors, and even if not, you can annotate them. FastAPI has long used such mechanisms to build upon FastAPI (of course, I am not a maintainer like yourself, but I am an experienced FastAPI user speaking from the limited experience I have).
In any case, can you reconsider if this really should be implemented in the Pydnatic side first?

**Reply by YuriiMotov** [reply]
I admit I may be wrong, but for now I still think it should be done on Pydantic's side first.

Could you please explain how you see the process of request validation and response serialization with these new parameters (`read_only`, `write_only`)?

Let's start from `read_only`.
If `read_only=True` in your example is just kwargs (extra) for Pydantic, it will not affect validation. Thus the `uid` field will be required and validation will be failed if you don't pass it in the input payload.
How do you expect FastAPI should handle this?
I have my assumptions, but I want to hear you ideas first

**Reply by avivgood** [reply]
This is just pseudo code but I was thinking about something like this:

```

from pydantic import BaseModel

# AFAIK Pydantic 2 removed extra args but similar behaviour works by another means such as annotations, the json_schema attributes and more
class User(BaseModel):
  name: str 
  uid: str = Field(request=False, response=True)

# ok
@app.get("/user")
def get_user() -> User:
  pass

# error
@app.post("/users")
def create_user(user: User):
  pass

# FastAPI Error (when used as a FastAPI request body): uid field decleared as response only but no defualt value or default factory are provided

class User(BaseModel):
  name: str 
  uid: str = Field('', request=False, response=True)

# ok
@app.get("/user")
def get_user() -> User:
  pass

# ok
@app.post("/users")
def create_user(user: User):
  pass


# OpenAPI shows a single model for both use cases with a read only field uid

# POST /users

# {name: "test", uid: "124"}

# 422: field uid is read only
```

**Reply by YuriiMotov** [reply]
I thought it over one more time.
Although this is possible to implement, I still believe this shouldn't be done in FastAPI (we should ask for this to be implemented in Pydantic instead).
I think this logic is not as simple as it seems to be. And nobody wants to bring more bugs into FastAPI :)

But I'm not a person who can make such decision.
Let's wait for other people to speak up.

**Reply by avivgood** [reply]
@tiangolo what do you think?

**Reply by avivgood** [reply]
Hi, still waiting on updates

**Comment by yakubka** [comment]
Pydantic v2 supports `Field(frozen=True)` for read-only semantics and you can add `readOnly` or `writeOnly` directly via `json_schema_extra`.

```python
from pydantic import BaseModel, Field

class Item(BaseModel):
    id: int = Field(json_schema_extra={"readOnly": True})
    name: str
    secret: str = Field(json_schema_extra={"writeOnly": True})
```

Swagger UI reads the `readOnly` and `writeOnly` keywords from the JSON Schema and renders them accordingly: read-only fields are shown in the response schema but greyed out in the request form, and write-only fields are shown in the request form but omitted from the response schema.

You can also split into separate request and response models, which is the more explicit and commonly recommended pattern.

```python
class ItemCreate(BaseModel):   # used for POST body
    name: str
    secret: str

class ItemResponse(BaseModel):  # used for response_model
    id: int
    name: str
    # secret omitted
```

Separate models make the distinction explicit in code and in the generated spec without relying on `readOnly`/`writeOnly` keywords.
